src/modules/hip/kernel/normalize.hpp

```diff
@@ -0,0 +1,1925 @@
+        for(uint i = 0; i < lengthZ; i++)
+        {
+            accum += srcPtr[srcIdx];
+            srcIdx += srcStridesNZY.y;
+        }
+        meanTensor[dstIdx] = accum / static_cast<float>(lengthZ);
+    }
+    // compute mean along y direction
+    else if(axisMask == 2)
+    {
+        if(id_x >= lengthX || id_y >= lengthZ)
+            return;
+
+        uint srcIdx = id_z * srcStridesNZY.x + id_y * srcStridesNZY.y + id_x;
+        uint dstIdx = id_z * maxParamVolume +  id_y * lengthX + id_x;
+        float accum = 0.0f;
+        for(uint i = 0; i < lengthY; i++)
+        {
+            accum += srcPtr[srcIdx];
+            srcIdx += srcStridesNZY.z;
+        }
+        meanTensor[dstIdx] = accum / static_cast<float>(lengthY);
+    }
+    // compute mean along x direction
+    else if(axisMask == 4)
+    {
+        if(id_x >= lengthY || id_y >= lengthZ)
+            return;
+
+        uint srcIdx = id_z * srcStridesNZY.x + id_y * srcStridesNZY.y + id_x * srcStridesNZY.z;
+        d_float8 accum_f8;
+        accum_f8.f4[0] = (float4)0.0f;
+        accum_f8.f4[1] = (float4)0.0f;
+        for(int i = 0; i < lengthX; i += 8)
+        {
+            d_float8 src_f8;
+            rpp_hip_load8_and_unpack_to_float8(srcPtr + srcIdx, &src_f8);
+            if (i + 8 > lengthX)
+            {
+                int xDiff = i + 8 - lengthX;
+                for(int i = xDiff; i < 8; i++)
+                    src_f8.f1[i] = 0.0f;
+            }
+            accum_f8.f4[0] += src_f8.f4[0];
+            accum_f8.f4[1] += src_f8.f4[1];
+            srcIdx += 8;
+        }
+        accum_f8.f4[0] += accum_f8.f4[1];
+        accum_f8.f1[0] = (accum_f8.f1[0] + accum_f8.f1[1] + accum_f8.f1[2] + accum_f8.f1[3]);
+        uint dstIdx = id_z * maxParamVolume + id_y * lengthY + id_x;
+        meanTensor[dstIdx] = accum_f8.f1[0] / static_cast<float>(lengthX);
+    }
+    // compute mean along x-z direction
+    else if(axisMask == 5)
+    {
+        __shared__ float partialSum_smem[32];
+        partialSum_smem[hipThreadIdx_x] = 0.0f;
+
+        if(hipBlockIdx_x >= lengthY)
+            return;
+
+        uint dstIdx = id_z * maxParamVolume +  hipBlockIdx_x;
+        float accum = 0.0f;
+        for (uint i = 0; i < lengthZ; i++)
+        {
+            uint tid_x = hipThreadIdx_x;
+            uint srcIdx = (id_z * srcStridesNZY.x) + (i * srcStridesNZY.y) + (hipBlockIdx_x * srcStridesNZY.z);
+            while (tid_x < lengthX)
+            {
+                accum += srcPtr[srcIdx + tid_x];
+                tid_x += hipBlockDim_x;
+            }
+        }
+        partialSum_smem[hipThreadIdx_x] = accum;
+        __syncthreads();
+
+        // perform reduction on shared memory sums
+        reduction_sum_x_hip(partialSum_smem);
+
+        if(hipThreadIdx_x == 0)
+            meanTensor[dstIdx] = partialSum_smem[0] / static_cast<float>(lengthX * lengthZ);
+    }
+    // compute mean along x-y-z direction
+    else if(axisMask == 7)
```

Comment:
Any specific reason for the ordering of the axisMask in the code as 1, 2, 4, 5, 7, 6, 3?

Reply:
Modified the order. The original thought was to separate 1 axis Normalization, 2 axis Normalization, 3 axis Normalization. The order now is as {1, 2, 4, 3, 6, 5, 7}
for 3D
axisMask 1 - compute mean along Z direction
axisMask 2 - compute mean along Y direction
axisMask 4 - compute mean along X direction
axisMask 3 - compute mean along Z - Y direction
axisMask 6 - compute mean along Y - X direction
axisMask 5 - compute mean along Z - X direction
axisMask 7 - compute mean along Z - Y - X direction